File: lib/features/intelligence/domain/entities/result_provenance.dart
/// Phase 0 — machine-readable result provenance (mirrors API contract).
enum ResultSource {
  providerMeasured,
  localMeasured,
  locallyCalculated,
  inferred,
  heuristic,
  userSupplied,
  mock,
  unavailable,
}

enum ProvenanceConfidenceLevel {
  high,
  medium,
  low,
  unavailable,
}

class ResultProvenance {
  final ResultSource resultSource;
  final String provider;
  final String? providerVersion;
  final String calculationVersion;
  final int confidence;
  final ProvenanceConfidenceLevel confidenceLevel;
  final String? captureQuality;
  final DateTime generatedAt;
  final List<String> limitations;
  final bool isMock;
  final bool canDisplay;
  final String? unavailableReason;

  const ResultProvenance({
    required this.resultSource,
    required this.provider,
    this.providerVersion,
    required this.calculationVersion,
    required this.confidence,
    required this.confidenceLevel,
    this.captureQuality,
    required this.generatedAt,
    required this.limitations,
    required this.isMock,
    required this.canDisplay,
    this.unavailableReason,
  });

  static ResultProvenance? tryParse(dynamic raw) {
    if (raw is! Map) return null;
    final map = Map<String, dynamic>.from(raw);
    return ResultProvenance(
      resultSource: _source(map['resultSource'] as String?),
      provider: map['provider'] as String? ?? 'unknown',
      providerVersion: map['providerVersion'] as String?,
      calculationVersion: map['calculationVersion'] as String? ?? 'unknown',
      confidence: (map['confidence'] as num?)?.toInt() ?? 0,
      confidenceLevel: _level(map['confidenceLevel'] as String?),
      captureQuality: map['captureQuality'] as String?,
      generatedAt: DateTime.tryParse(map['generatedAt'] as String? ?? '') ??
          DateTime.fromMillisecondsSinceEpoch(0),
      limitations: (map['limitations'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
      isMock: map['isMock'] as bool? ?? false,
      canDisplay: map['canDisplay'] as bool? ?? true,
      unavailableReason: map['unavailableReason'] as String?,
    );
  }

  static ResultSource _source(String? v) => switch (v) {
        'provider_measured' => ResultSource.providerMeasured,
        'local_measured' => ResultSource.localMeasured,
        'locally_calculated' => ResultSource.locallyCalculated,
        'inferred' => ResultSource.inferred,
        'heuristic' => ResultSource.heuristic,
        'user_supplied' => ResultSource.userSupplied,
        'mock' => ResultSource.mock,
        _ => ResultSource.unavailable,
      };

  static ProvenanceConfidenceLevel _level(String? v) => switch (v) {
        'high' => ProvenanceConfidenceLevel.high,
        'medium' => ProvenanceConfidenceLevel.medium,
        'low' => ProvenanceConfidenceLevel.low,
        _ => ProvenanceConfidenceLevel.unavailable,
      };
}

abstract final class CosmeticCopy {
  CosmeticCopy._();

  static const disclaimerAr =
      'هذا التحليل تجميلي وإرشادي، وليس تشخيصاً طبياً. قد تختلف النتائج باختلاف الإضاءة والكاميرا وجودة الصورة.';

  static const disclaimerEn =
      'This analysis is cosmetic and informational, not a medical diagnosis. Results may vary based on lighting, camera and image quality.';

  static const skinVitalityIndexAr = 'مؤشر حيوية البشرة';
  static const skinVitalityIndexEn = 'Skin Vitality Index';

  static const skinVitalitySupportingAr =
      'تقدير تجميلي مبني على مؤشرات البشرة الظاهرة في الصورة، وقد تتأثر النتيجة بالإضاءة وجودة الكاميرا.';
}

File: mira-api/src/intelligence/contracts/result-provenance.ts
/**
 * Phase 0 — Result provenance & confidence (machine-readable truth labels).
 */

export type ResultSource =
  | 'provider_measured'
  | 'local_measured'
  | 'locally_calculated'
  | 'inferred'
  | 'heuristic'
  | 'user_supplied'
  | 'mock'
  | 'unavailable';

export type ConfidenceLevel = 'high' | 'medium' | 'low' | 'unavailable';

export interface ResultProvenance {
  resultSource: ResultSource;
  provider: string;
  providerVersion?: string;
  calculationVersion: string;
  confidence: number;
  confidenceLevel: ConfidenceLevel;
  captureQuality?: string;
  generatedAt: string;
  limitations: string[];
  isMock: boolean;
  canDisplay: boolean;
  unavailableReason?: string;
}

/** Phase 3 — Skin Vitality Index formula family. */
export const SKIN_VITALITY_CALCULATION_VERSION = 'svi-v2';
export const SCORE_SCHEMA_VERSION = 2;

export const DISCLAIMER_AR =
  'هذا التحليل تجميلي وإرشادي، وليس تشخيصاً طبياً. قد تختلف النتائج باختلاف الإضاءة والكاميرا وجودة الصورة.';

export const DISCLAIMER_EN =
  'This analysis is cosmetic and informational, not a medical diagnosis. Results may vary based on lighting, camera and image quality.';

export const SKIN_VITALITY_LABEL_AR = 'مؤشر حيوية البشرة';
export const SKIN_VITALITY_LABEL_EN = 'Skin Vitality Index';

export const SKIN_VITALITY_SUPPORTING_AR =
  'تقدير تجميلي مبني على مؤشرات البشرة الظاهرة في الصورة، وقد تتأثر النتيجة بالإضاءة وجودة الكاميرا.';

export function confidenceLevelFromScore(confidence: number): ConfidenceLevel {
  if (!Number.isFinite(confidence) || confidence <= 0) return 'unavailable';
  if (confidence >= 80) return 'high';
  if (confidence >= 55) return 'medium';
  return 'low';
}

export function buildSkinVitalityProvenance(input: {
  isMock: boolean;
  provider: string;
  providerVersion?: string;
  confidence: number;
  captureQuality?: string;
  limitations?: string[];
  isProduction: boolean;
}): ResultProvenance {
  const isMock = input.isMock === true;
  const canDisplay = !(isMock && input.isProduction);
  return {
    resultSource: isMock ? 'mock' : 'locally_calculated',
    provider: input.provider,
    providerVersion: input.providerVersion,
    calculationVersion: SKIN_VITALITY_CALCULATION_VERSION,
    confidence: Math.round(Math.min(100, Math.max(0, input.confidence))),
    confidenceLevel: isMock
      ? 'unavailable'
      : confidenceLevelFromScore(input.confidence),
    captureQuality: input.captureQuality,
    generatedAt: new Date().toISOString(),
    limitations: [
      ...(input.limitations ?? []),
      SKIN_VITALITY_SUPPORTING_AR,
      DISCLAIMER_AR,
    ],
    isMock,
    canDisplay,
    unavailableReason: canDisplay
      ? undefined
      : 'Mock results cannot be displayed in production',
  };
}

/** UI / API gate: never show mock as genuine in production. */
export function assertDisplayableInProduction(
  provenance: ResultProvenance,
  isProduction: boolean,
): void {
  if (isProduction && (provenance.isMock || !provenance.canDisplay)) {
    throw new Error(
      provenance.unavailableReason ??
        'Result marked mock/unavailable and cannot be displayed in production',
    );
  }
}
